lib/shelley/src/Cardano/Wallet/Shelley/Network/Node.hs

```diff
@@ -1043,70 +1001,52 @@ recoveringNodeConnection tr handlers action =
-        pure DontRetry
-  where
-    isResourceVanishedError = isInfixOf "resource vanished" . show
-    isTryAgainError = isInfixOf "resource exhausted" . show
-
-handleMuxError
-    :: Tracer IO (Maybe IOException)
-    -> Bool -- ^ 'True' = retry on 'ResourceVanishedError'
-    -> MuxError
-    -> IO RetryAction
-handleMuxError tr onResourceVanished = pure . errorType >=> \case
-    MuxUnknownMiniProtocol -> pure DontRetry
-    MuxDecodeError -> pure DontRetry
-    MuxIngressQueueOverRun -> pure DontRetry
-    MuxInitiatorOnly -> pure DontRetry
-    MuxShutdown _ -> pure DontRetry -- fixme: #2212 consider cases
-    MuxCleanShutdown -> pure DontRetry
-    MuxIOException e ->
-        handleIOException tr onResourceVanished e
-    MuxBearerClosed -> do
-        traceWith tr Nothing
-        pure $ if onResourceVanished then ConsultPolicy else DontRetry
-
-    -- MuxSDU*Timeout errors arise because the bandwidth of the
-    -- interprocess communication socket dropped unexpectedly,
-    -- and the socket library decided to cut off the connection
-    -- after ~ 30 seconds.
-    -- Chances are that the system is overloaded, let's retry in 30 seconds.
-    MuxSDUReadTimeout -> pure $ ConsultPolicyOverrideDelay 30_000_000
-    MuxSDUWriteTimeout -> pure $ ConsultPolicyOverrideDelay 30_000_000
+handleIOException :: Tracer IO IOException -> IOException -> IO RetryAction
+handleIOException tr ioException =
+    traceWith tr ioException $> case ioException of
+        -- There's a race-condition when starting the wallet and the node at the
+        -- same time: the socket might not be there yet when we try to open it.
+        -- In such case, we simply retry a bit later and hope it's there.
+        _ | isDoesNotExistError ioException -> ConsultPolicy
+        -- I/O error where the operation failed because the resource vanished.
+        -- This happens when, for example, attempting to write to a closed
+        -- socket or attempting to write to a named pipe that was deleted.
+        e | isResourceVanishedError e -> ConsultPolicy
+        -- If the nonblocking UNIX domain socket connection cannot be completed
+        -- immediately (i.e. connect() returns EAGAIN), try again. This happens
+        -- because the node's listen queue is quite short.
+        e | "resource exhausted" `isInfixOf` show e -> ConsultPolicy
+        _ -> DontRetry
+
+handleMuxError :: Tracer IO ReasonConnectionLost -> MuxError -> IO RetryAction
+handleMuxError tr muxErr = do
+    traceWith tr (ReasonMuxError muxErr)
+    case errorType muxErr of
+        MuxUnknownMiniProtocol -> pure DontRetry
+        MuxDecodeError -> pure DontRetry
+        MuxIngressQueueOverRun -> pure DontRetry
+        MuxInitiatorOnly -> pure DontRetry
+        MuxShutdown _ -> pure DontRetry -- fixme: #2212 consider cases
+        MuxCleanShutdown -> pure DontRetry
+        MuxIOException e -> handleIOException (ReasonException >$< tr) e
+        MuxBearerClosed -> pure ConsultPolicy
```

Comment:
```suggestion
        MuxBearerClosed -> pure $ if onResourceVanished then ConsultPolicy else DontRetry
```

Please be very careful about not changing the semantics here — this stuff is trial & error, I remember there was a bug/issue associated with this line specifically.

Generally, I find that cosmetic changes on working code are worth the effort only in half of the cases — the cost in review time and the risk of introducing bugs often outweigh the benefit for code whose readability is "good enough".

Reply:
This function is only applied with `onResourceVanished == True`, so `DontRetry` is never evaluated.

Reply:
😱 You're right. Thank you! (Proof as supplied by @Unisay : https://github.com/input-output-hk/cardano-wallet/search?l=Haskell&q=handleIOException )